src/types.jl

```diff
@@ -35,22 +35,49 @@ mutable struct CoordTransform
     ptr::GDAL.OGRCoordinateTransformationH
 end
 
+# In the multidim API, underlying files are closed only when all
+# objects potentially pointing to them (groups, arrays, attributes,
+# dimensions) have been released. Their lifetime is not connected with
+# the one of the GDALDataset.
+#
+# To handle this in Julia, each multidim dataset can hold a list of
+# its children. This allows us to "hard close" a dataset when using
+# the interactive API. Being able to close a dataset at a particular
+# time is important when writing.
+#
+# Each child must have a `destroy` function.
 mutable struct Dataset <: AbstractDataset
     ptr::GDAL.GDALDatasetH
+    children::Union{Nothing,Vector{WeakRef}}
 
-    Dataset(ptr::GDAL.GDALDatasetH = C_NULL) = new(ptr)
+    function Dataset(ptr::GDAL.GDALDatasetH = C_NULL; hard_close::Bool = false)
+        return new(ptr, hard_close ? WeakRef[] : nothing)
+    end
 end
 
 mutable struct IDataset <: AbstractDataset
     ptr::GDAL.GDALDatasetH
+    children::Union{Nothing,Vector{WeakRef}}
 
-    function IDataset(ptr::GDAL.GDALDatasetH = C_NULL)
-        dataset = new(ptr)
+    function IDataset(ptr::GDAL.GDALDatasetH = C_NULL; hard_close::Bool = false)
+        dataset = new(ptr, hard_close ? WeakRef[] : nothing)
         finalizer(destroy, dataset)
         return dataset
     end
 end
 
+function add_child!(dataset::WeakRef, obj::Any)::Nothing
```

Comment:
I'd recommend renaming it to `_add_mdarray_child!` moving this function to `src/mdarray/types.jl` since that's where its used (and we haven't verified that it's something we'll want to be supporting for the rest of GDAL's data types)